File: exercises/day10-TimePortal/java/src/test/java/timeportal/TravelerTest.java
package timeportal;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("Traveler")
class TravelerTest {

    private Traveler traveler;

    @BeforeEach
    void setUp() {
        traveler = new Traveler("Dr. Elena Vance", "VISA-2024-001");
    }

    @Nested
    @DisplayName("Initial State")
    class InitialState {

        @Test
        @DisplayName("should have correct name")
        void shouldHaveCorrectName() {
            assertEquals("Dr. Elena Vance", traveler.getName());
        }

        @Test
        @DisplayName("should have correct visa code")
        void shouldHaveCorrectVisaCode() {
            assertEquals("VISA-2024-001", traveler.getVisaCode());
        }

        @Test
        @DisplayName("should start with zero jumps")
        void shouldStartWithZeroJumps() {
            assertEquals(0, traveler.getJumpCount());
        }

        @Test
        @DisplayName("should start with STABLE health")
        void shouldStartWithStableHealth() {
            assertEquals("STABLE", traveler.getHealthStatus());
        }

        @Test
        @DisplayName("should not have temporal sickness initially")
        void shouldNotHaveTemporalSicknessInitially() {
            assertFalse(traveler.hasTemporalSickness());
        }
    }

    @Nested
    @DisplayName("Year Tracking")
    class YearTracking {

        @Test
        @DisplayName("should track current year")
        void shouldTrackCurrentYear() {
            traveler.setCurrentYear(1985);
            assertEquals(1985, traveler.getCurrentYear());
        }
    }

    @Nested
    @DisplayName("Jump Count and Health")
    class JumpCountAndHealth {

        @Test
        @DisplayName("should increment jump count")
        void shouldIncrementJumpCount() {
            traveler.incrementJumpCount();
            assertEquals(1, traveler.getJumpCount());
        }

        @Test
        @DisplayName("should degrade health after many jumps")
        void shouldDegradeHealthAfterManyJumps() {
            for (int i = 0; i < 6; i++) {
                traveler.incrementJumpCount();
            }
            assertNotEquals("STABLE", traveler.getHealthStatus());
        }

        @Test
        @DisplayName("should become critical after excessive jumps")
        void shouldBecomeCriticalAfterExcessiveJumps() {
            for (int i = 0; i < 11; i++) {
                traveler.incrementJumpCount();
            }
            assertEquals("CRITICAL", traveler.getHealthStatus());
        }
    }

    @Nested
    @DisplayName("Temporal Sickness")
    class TemporalSickness {

        @Test
        @DisplayName("should have temporal sickness after excessive jumps")
        void shouldHaveTemporalSicknessAfterManyJumps() {
            for (int i = 0; i < 11; i++) {
                traveler.incrementJumpCount();
            }
            assertTrue(traveler.hasTemporalSickness());
        }

        @Test
        @DisplayName("should have temporal sickness immediately after a jump")
        void shouldHaveTemporalSicknessRightAfterJump() {
            traveler.incrementJumpCount();
            assertTrue(traveler.hasTemporalSickness());
        }
    }
}

File: exercises/day10-TimePortal/csharp/TimePortal/TimePortal.Tests/Behavior/TravelerTest.cs
using Xunit;

namespace TimePortal.Tests;

public class TravelerTest
{
    private readonly Traveler _traveler;

    public TravelerTest()
    {
        _traveler = new Traveler("Dr. Elena Vance", "VISA-2024-001");
    }

    #region Initial State

    [Fact(DisplayName = "Should have correct name")]
    public void ShouldHaveCorrectName()
    {
        Assert.Equal("Dr. Elena Vance", _traveler.Name);
    }

    [Fact(DisplayName = "Should have correct visa code")]
    public void ShouldHaveCorrectVisaCode()
    {
        Assert.Equal("VISA-2024-001", _traveler.VisaCode);
    }

    [Fact(DisplayName = "Should start with zero jumps")]
    public void ShouldStartWithZeroJumps()
    {
        Assert.Equal(0, _traveler.JumpCount);
    }

    [Fact(DisplayName = "Should start with STABLE health")]
    public void ShouldStartWithStableHealth()
    {
        Assert.Equal("STABLE", _traveler.GetHealthStatus());
    }

    [Fact(DisplayName = "Should not have temporal sickness initially")]
    public void ShouldNotHaveTemporalSicknessInitially()
    {
        Assert.False(_traveler.HasTemporalSickness());
    }

    #endregion

    #region Year Tracking

    [Fact(DisplayName = "Should track current year")]
    public void ShouldTrackCurrentYear()
    {
        _traveler.SetCurrentYear(1985);
        Assert.Equal(1985, _traveler.CurrentYear);
    }

    #endregion

    #region Jump Count and Health

    [Fact(DisplayName = "Should increment jump count")]
    public void ShouldIncrementJumpCount()
    {
        _traveler.IncrementJumpCount();
        Assert.Equal(1, _traveler.JumpCount);
    }

    [Fact(DisplayName = "Should degrade health after many jumps")]
    public void ShouldDegradeHealthAfterManyJumps()
    {
        for (var i = 0; i < 6; i++)
        {
            _traveler.IncrementJumpCount();
        }
        Assert.NotEqual("STABLE", _traveler.GetHealthStatus());
    }

    [Fact(DisplayName = "Should become critical after excessive jumps")]
    public void ShouldBecomeCriticalAfterExcessiveJumps()
    {
        for (var i = 0; i < 11; i++)
        {
            _traveler.IncrementJumpCount();
        }
        Assert.Equal("CRITICAL", _traveler.GetHealthStatus());
    }

    #endregion

    #region Temporal Sickness

    [Fact(DisplayName = "Should have temporal sickness after excessive jumps")]
    public void ShouldHaveTemporalSicknessAfterManyJumps()
    {
        for (var i = 0; i < 11; i++)
        {
            _traveler.IncrementJumpCount();
        }
        Assert.True(_traveler.HasTemporalSickness());
    }

    [Fact(DisplayName = "Should have temporal sickness immediately after a jump")]
    public void ShouldHaveTemporalSicknessRightAfterJump()
    {
        _traveler.IncrementJumpCount();
        Assert.True(_traveler.HasTemporalSickness());
    }

    #endregion
}
